lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2171,16 +2149,93 @@ SynthesizeMainFunctionRequest::evaluate(Evaluator &evaluator,
+  // `() throws -> Void`
+  // `() async throws -> Void`
+  // `@MainActor () -> Void`
+  // `@MainActor () async -> Void`
+  // `@MainActor () throws -> Void`
+  // `@MainActor () async throws -> Void`
+  {
+    llvm::SmallVector<Type, 8> mainTypes = {
+
+        FunctionType::get(/*params*/ {}, context.TheEmptyTupleType,
+                          ASTExtInfo()),
+        FunctionType::get(
+            /*params*/ {}, context.TheEmptyTupleType,
+            ASTExtInfoBuilder().withAsync().build()),
+
+        FunctionType::get(/*params*/ {}, context.TheEmptyTupleType,
+                          ASTExtInfoBuilder().withThrows().build()),
+
+        FunctionType::get(
+            /*params*/ {}, context.TheEmptyTupleType,
+            ASTExtInfoBuilder().withAsync().withThrows().build())};
+
+    Type mainActor = context.getMainActorType();
+    if (mainActor) {
+      mainTypes.push_back(FunctionType::get(
+          /*params*/ {}, context.TheEmptyTupleType,
+          ASTExtInfoBuilder().withGlobalActor(mainActor).build()));
+      mainTypes.push_back(FunctionType::get(
+          /*params*/ {}, context.TheEmptyTupleType,
+          ASTExtInfoBuilder().withAsync().withGlobalActor(mainActor).build()));
+      mainTypes.push_back(FunctionType::get(
+          /*params*/ {}, context.TheEmptyTupleType,
+          ASTExtInfoBuilder().withThrows().withGlobalActor(mainActor).build()));
+      mainTypes.push_back(FunctionType::get(/*params*/ {},
+                                            context.TheEmptyTupleType,
+                                            ASTExtInfoBuilder()
+                                                .withAsync()
+                                                .withThrows()
+                                                .withGlobalActor(mainActor)
+                                                .build()));
+    }
+    TypeVariableType *mainType =
+        CS.createTypeVariable(CS.getConstraintLocator({}), /*options=*/0);
+    llvm::SmallVector<constraints::Constraint *, 4> typeEqualityConstraints;
+    typeEqualityConstraints.reserve(mainTypes.size());
+    for (const Type &candidateMainType : mainTypes) {
+      constraints::Constraint *fnEqualConstraint =
+          constraints::Constraint::create(
+              CS, constraints::ConstraintKind::Equal, Type(mainType),
+              candidateMainType, locator);
+      typeEqualityConstraints.push_back(fnEqualConstraint);
+    }
+
+    CS.addDisjunctionConstraint(typeEqualityConstraints, locator);
+    CS.addValueMemberConstraint(
+        nominal->getInterfaceType(), DeclNameRef(context.Id_main),
+        Type(mainType), declContext, FunctionRefKind::SingleApply, {}, locator);
+  }
+
+  FuncDecl *mainFunction = nullptr;
+  llvm::SmallVector<constraints::Solution, 4> candidates;
+
+  if (!CS.solve(candidates, FreeTypeVariableBinding::Disallow)) {
+    CS.diagnoseAmbiguity(candidates);
+    // We can't use CS.diagnoseAmbiguity directly since the locator is empty
```

Comment:
Have you tried using `nominal` as an anchor for `locator`?

Reply:
Yeah, I still hit the `assert(false && "locator could not be simplified to anchor")`: https://github.com/apple/swift/blob/c891f34b22d93432016ff544f17c591fda26abcb/lib/Sema/ConstraintSystem.cpp#L4770-L4780

Reply:
Alright, something we need to improve then.